qutip/krylovsolve.py

```diff
@@ -0,0 +1,632 @@
+        krylov_results.expect = {
+            e: krylov_results.expect[n]
+            for n, e in enumerate(e_ops_dict.keys())
+        }
+
+    return krylov_results
+
+
+def _expectation_values(
+    e_ops,
+    n_expt_op,
+    expt_callback,
+    res,
+    evolved_states,
+    partitions,
+    idx,
+    t_idx,
+    options,
+):
+
+    if options.store_states:
+        res.states += evolved_states
+
+    for t, state in zip(
+        range(t_idx, t_idx + len(partitions[idx][1:-1])), evolved_states
+    ):
+
+        if expt_callback:
+            # use callback method
+            res.expect.append(e_ops(t, state))
+
+        for m in range(n_expt_op):
+            op = e_ops[m]
+            if not isinstance(op, Qobj) and callable(op):
+                res.expect[m][t] = op(t, state)
+                continue
+
+            res.expect[m][t] = expect(op, state)
+
+    if idx == len(partitions) - 1:
+        if (options.store_final_state) and (not options.store_states):
+            res.states = [evolved_states[-1]]
+
+    return res
+
+
+def lanczos_algorithm(
+    H,
+    psi: np.ndarray,
+    krylov_dim: int,
+    sparse: bool = False,
+):
+    """
+    Computes a basis of the Krylov subspace for Hamiltonian 'H', a system
+    state 'psi' and Krylov dimension 'krylov_dim'. The space is spanned
+    by {psi, H psi, H^2 psi, ..., H^(krylov_dim) psi}.
+    Parameters
+    ------------
+    H : np.ndarray or csr_matrix
+       System Hamiltonian. If the Hamiltonian is dense, a np.ndarray is
+       preferred, whereas if it is sparse, a scipy csr_matrix is optimal.
+    psi: np.ndarray
+        State used to calculate Krylov subspace.
+    krylov_dim: int
+        Dimension (krylov_dim + 1) of the spanned Krylov subspace.
+    sparse: bool (optional, default False)
+        Wether to perform scipy sparse matrix multiplication operations or
+        numpy dense matrix multiplications.
+    Returns
+    ---------
+    v: np.ndarray
+        Lanczos eigenvector.
+    T: np.ndarray
+        Tridiagonal decomposition.
+    """
+
+    v = np.zeros((krylov_dim + 2, psi.shape[0]), dtype=complex)
+    T_m = np.zeros((krylov_dim + 2, krylov_dim + 2), dtype=complex)
+
+    v[0, :] = psi.squeeze()
+
+    w_prime = H.dot(v[0, :])
+
+    alpha = np.vdot(w_prime, v[0, :])
+
+    w = w_prime - alpha * v[0, :]
+
+    T_m[0, 0] = alpha
+
+    for j in range(1, krylov_dim + 1):
+
+        beta = np.linalg.norm(w)
+
+        if beta < 1e-7:
+
+            # Happy breakdown
+            v_happy = v[0:j, :]
+            T_m_happy = T_m[0:j, 0:j]
+
+            return v_happy, T_m_happy
+
+        v[j, :] = w / beta
+        w_prime = H.dot(v[j, :])
+        alpha = np.vdot(w_prime, v[j, :])
+
+        w = w_prime - alpha * v[j, :] - beta * v[j - 1, :]
+
+        T_m[j, j] = alpha
+        T_m[j, j - 1] = beta
+        T_m[j - 1, j] = beta
+
+    beta = np.linalg.norm(w)
+    v[krylov_dim + 1, :] = w / beta
+
+    T_m[krylov_dim + 1, krylov_dim] = beta
```

Comment:
Shouldn't this be followed by `T_m[krylov_dim, krylov_dim+1] = beta`?
Otherwise T is not Hermitian  and since the Hermitian eigen value solver is used (`eigh`) this can result in errors. This is the cause of the error with python 3.6.

Reply:
This is a really interesting finding. 

What you mention is correct regarding it being Hermitian. However, the correct approach right now is to remove those lines of code, because the documentation says <code>T.shape = [krylov_dim + 1 , krylov_dim + 1]</code>, instead of <code>[krylov_dim + 2, krylov_dim + 2]</code>.